File: frontend/src/utils/security.ts
/**
 * Content Security Policy (CSP) 配置
 * 用于防止 XSS、点击劫持等攻击
 *
 * 注意: 某些安全响应头只能通过 HTTP 响应头设置，不能通过 meta 标签设置:
 * - X-Frame-Options (必须在后端设置)
 * - X-Content-Type-Options (必须在后端设置)
 * - X-XSS-Protection (必须在后端设置)
 * - Strict-Transport-Security (必须在后端设置)
 *
 * 请参考 docs/SECURITY_HEADERS.md 了解如何在后端配置这些响应头
 */

// 动态获取 API URL
function getApiUrl(): string {
  // 优先使用环境变量（去除首尾空格）
  const envUrl = (import.meta.env.PUBLIC_API_URL || '').trim()
  if (envUrl) {
    return envUrl
  }

  // 浏览器环境：使用当前域名（空字符串表示相对路径）
  if (typeof window !== 'undefined') {
    const origin = window.location.origin
    // 如果是标准端口（生产环境），返回空字符串使用相对路径
    if (window.location.port === '' || window.location.port === '80' || window.location.port === '443') {
      return ''
    }
    // 开发环境返回当前 origin
    return origin
  }

  // SSR/构建时默认值：空字符串（使用相对路径）
  return ''
}

// 动态生成 connect-src 列表
function getConnectSources(): string[] {
  const sources = ['\'self\'']

  const apiUrl = getApiUrl()

  // 添加 API URL（如果不是 'self'）
  if (apiUrl && apiUrl !== '') {
    try {
      const url = new URL(apiUrl)
      const apiOrigin = url.origin
      if (apiOrigin !== (typeof window !== 'undefined' ? window.location.origin : '')) {
        sources.push(apiOrigin)
      }
    }
    catch {
      // 如果解析失败，尝试直接添加
      if (apiUrl.startsWith('http')) {
        sources.push(apiUrl)
      }
    }
  }

  // 添加常见的本地开发地址（仅开发环境）
  if (typeof window !== 'undefined' && (
    window.location.hostname === 'localhost'
    || window.location.hostname === '127.0.0.1'
  )) {
    // 开发环境才添加 localhost 地址
    if (window.location.port !== '' && window.location.port !== '80' && window.location.port !== '443') {
      sources.push('http://localhost:3000')
      sources.push('http://127.0.0.1:3000')
    }
  }

  // 添加其他必需的外部服务
  sources.push('https://api.github.com')
  sources.push('https://image.pollinations.ai')

  // 去重
  return Array.from(new Set(sources))
}

export const CSP_DIRECTIVES = {
  // 默认源：只允许同源内容
  'default-src': ['\'self\''],

  // 脚本源：允许同源和内联脚本（Astro需要）
  'script-src': [
    '\'self\'',
    '\'unsafe-inline\'', // Astro 内联脚本需要
    '\'unsafe-eval\'', // 开发环境需要，生产环境应移除
  ],

  // 样式源：允许同源和内联样式
  'style-src': [
    '\'self\'',
    '\'unsafe-inline\'', // Tailwind CSS 需要
    'https://fonts.googleapis.com',
  ],

  // 字体源
  'font-src': [
    '\'self\'',
    'https://fonts.gstatic.com',
  ],

  // 图片源:允许同源、data URI 和外部图片服务
  'img-src': [
    '\'self\'',
    'data:',
    'blob:',
    'https:', // 允许所有 HTTPS 图片(壁纸服务)
  ],

  // 媒体源
  'media-src': ['\'self\''],

  // 连接源：API 请求 - 动态生成
  'connect-src': [], // 将在 generateCSPString 中动态填充

  // Frame 源：禁止嵌入
  'frame-src': ['\'none\''],

  // Object 源：禁止插件
  'object-src': ['\'none\''],

  // Base URI：限制 <base> 标签
  'base-uri': ['\'self\''],

  // Form 动作：限制表单提交
  'form-action': ['\'self\''],

  // Frame 祖先：防止点击劫持
  'frame-ancestors': ['\'none\''],

  // 升级不安全请求（生产环境）
  'upgrade-insecure-requests': [],
}

/**
 * 生成 CSP 字符串
 */
export function generateCSPString(isDev: boolean = false): string {
  const directives: Record<string, string[]> = { ...CSP_DIRECTIVES }

  // 动态设置 connect-src
  directives['connect-src'] = getConnectSources()

  // 生产环境移除 unsafe-eval
  if (!isDev && directives['script-src']) {
    directives['script-src'] = directives['script-src'].filter(
      src => src !== '\'unsafe-eval\'',
    )
  }

  return Object.entries(directives)
    .map(([key, values]) => {
      if (values.length === 0)
        return key
      return `${key} ${values.join(' ')}`
    })
    .join('; ')
}

/**
 * 其他安全响应头配置
 *
 * ⚠️ 注意: 这些响应头必须在后端设置，不能通过 HTML meta 标签设置
 * 请在 Rust 后端或 Nginx 反向代理中配置这些响应头
 */
export const SECURITY_HEADERS = {
  // 防止点击劫持 (必须在后端设置)
  'X-Frame-Options': 'DENY',

  // 防止 MIME 类型嗅探 (必须在后端设置)
  'X-Content-Type-Options': 'nosniff',

  // XSS 保护 - 旧浏览器 (必须在后端设置)
  'X-XSS-Protection': '1; mode=block',

  // Referrer 策略 (可以通过 meta 标签设置，已在 Layout.astro 中配置)
  'Referrer-Policy': 'strict-origin-when-cross-origin',

  // 权限策略 (必须在后端设置)
  'Permissions-Policy': 'geolocation=(), microphone=(), camera=()',

  // HSTS - 仅生产环境且使用 HTTPS 时启用 (必须在后端设置)
  'Strict-Transport-Security': 'max-age=31536000; includeSubDomains; preload',
}

/**
 * 将安全头转换为 meta 标签
 *
 * ⚠️ 注意: 只有 CSP 和 Referrer-Policy 可以通过 meta 标签设置
 * 其他安全响应头必须在后端设置
 */
export function generateSecurityMetaTags(isDev: boolean = false): string {
  const csp = generateCSPString(isDev)

  return `
    <meta name="referrer" content="strict-origin-when-cross-origin">
    ${!isDev ? `<meta http-equiv="Content-Security-Policy" content="${csp}">` : ''}
  `.trim()
}

File: backend/src/middleware/security.rs
use axum::{extract::Request, http::header, middleware::Next, response::Response};
use std::env;

/// Security headers middleware - adds CSP and other security headers
pub async fn security_headers_middleware(req: Request, next: Next) -> Response {
    let mut response = next.run(req).await;
    let headers = response.headers_mut();

    // Content Security Policy (CSP)
    let is_production =
        env::var("ENVIRONMENT").unwrap_or_else(|_| "development".to_string()) == "production";

    // img-src: dual-path image loading (MYR-039 / image proxy).
    // Non-hotlink https hosts are used as original URLs in <img src>, so CSP must
    // allow http(s) remote images. We avoid the bare `*` scheme wildcard (which
    // would also permit data-adjacent exotic schemes) while keeping dual-path working.
    // See docs/deployment/SECURITY_HEADERS.md §CSP img-src dual-path.
    const IMG_SRC: &str = "img-src 'self' data: blob: https: http:";

    let csp = if is_production {
        // PRODUCTION: Get allowed API origins from env (fallback to default)
        // wss/stun/turn: Shengwang realtime talk (Agora RTC) plus any other WebRTC.
        let allowed_api_origins = env::var("CSP_CONNECT_SRC")
            .unwrap_or_else(|_| "'self' https: wss: stun: turn:".to_string());

        // PRODUCTION: Strict CSP for executable content, relaxed for assets
        format!(
            "default-src 'self'; \
            script-src 'self'; \
            style-src 'self' 'unsafe-inline' https:; \
            {IMG_SRC}; \
            font-src 'self' data: https: blob:; \
            media-src 'self' https: blob:; \
            connect-src {}; \
            object-src 'none'; \
            base-uri 'self'; \
            form-action 'self'; \
            frame-ancestors 'none'; \
            frame-src 'none'; \
            worker-src 'self' blob:; \
            manifest-src 'self'; \
            upgrade-insecure-requests; ",
            allowed_api_origins
        )
    } else {
        // DEVELOPMENT: Relaxed CSP for hot reload and dev tools
        format!(
            "default-src 'self'; \
            script-src 'self' 'unsafe-inline' 'unsafe-eval'; \
            style-src 'self' 'unsafe-inline' https:; \
            {IMG_SRC}; \
            font-src 'self' data: https: blob:; \
            media-src 'self' https: blob:; \
            connect-src 'self' ws: wss: https: http:; \
            object-src 'none'; \
            base-uri 'self'; \
            worker-src 'self' blob:; "
        )
    };

    if is_production || env::var("ENABLE_CSP_DEV").unwrap_or_default() == "true" {
        if let Ok(header_value) = csp.parse() {
            headers.insert(header::CONTENT_SECURITY_POLICY, header_value);
            tracing::debug!(
                "CSP enabled: {}",
                if is_production {
                    "production"
                } else {
                    "development"
                }
            );
        } else {
            tracing::error!("Failed to parse CSP header value");
        }
    } else {
        tracing::debug!("CSP disabled in development mode (set ENABLE_CSP_DEV=true to enable)");
    }

    // X-Content-Type-Options: 防止MIME类型嗅探
    headers.insert(header::X_CONTENT_TYPE_OPTIONS, "nosniff".parse().unwrap());

    // X-Frame-Options: 防止点击劫持
    headers.insert(header::X_FRAME_OPTIONS, "DENY".parse().unwrap());

    // X-XSS-Protection: XSS过滤器（虽然现代浏览器已不需要，但为了兼容性保留）
    headers.insert(
        "X-XSS-Protection".parse::<header::HeaderName>().unwrap(),
        "1; mode=block".parse().unwrap(),
    );

    // Referrer-Policy: 控制Referrer信息泄漏
    headers.insert(
        header::REFERRER_POLICY,
        "strict-origin-when-cross-origin".parse().unwrap(),
    );

    // Permissions-Policy: 本站可定位（天气）和用麦克风（听/说）；摄像头仍禁用。
    // 文档级策略由 proxy（补齐）+ Astro dev middleware 共同保证；此处覆盖 API 响应。
    // 字符串须与 proxy PERMISSIONS_POLICY / frontend DOCUMENT_PERMISSIONS_POLICY 保持一致。
    headers.insert(
        "Permissions-Policy".parse::<header::HeaderName>().unwrap(),
        "geolocation=(self), microphone=(self), camera=()"
            .parse()
            .unwrap(),
    );

    // Strict-Transport-Security: 强制HTTPS（仅生产环境）
    if is_production {
        headers.insert(
            header::STRICT_TRANSPORT_SECURITY,
            "max-age=31536000; includeSubDomains".parse().unwrap(),
        );
    }

    response
}
